README.md
This library helps with CNPJ and CPF validations. Both of them are brazilian a documentation where CNPJ is a business document and CPF is a person document.

## Usage

A simple usage example:

```dart
import 'package:cnpj_cpf_helper/cnpj_cpf_helper.dart';

void main() {
  // generate a random cnpj
  var _cnpj = CnpjCpfBase.randomCnpj(withMask: true);
  print('a random CNPJ: $_cnpj');

  // masking a cnpj
  var _fromCnpj = '26954374000130';
  _cnpj = CnpjCpfBase.maskCnpj(_fromCnpj);
  print('masking a CNPJ, from this: $_fromCnpj to this: $_cnpj');
  // it's possible to masking partially
  _fromCnpj = '2695437';
  _cnpj = CnpjCpfBase.maskCnpj(_fromCnpj);
  print('masking partially a CNPJ, from this: $_fromCnpj to this: $_cnpj');

  // unmasking a cnpj
  _fromCnpj = '03.451.682/0001-47';
  _cnpj = CnpjCpfBase.onlyNumbers(_fromCnpj);
  print('unmasking  a CNPJ, from this: $_fromCnpj to this: $_cnpj');

  // validate a cnpj
  // there are two kinds of validation
  // the first one (the simplest one) just returns a bool value
  if (CnpjCpfBase.isCnpjValid(_cnpj)) {
    print('$_cnpj is a valid CNPJ');
  }

  // the second one returns a enum that describe the status
  _cnpj = '48.608.581/0005-00'; // invalid digit
  var _status = CnpjCpfBase.cnpjValidate(_cnpj);
  switch (_status) {
    case EDocumentStatus.DIGIT:
      print('CNPJ\'s digit is wrong!');
      break;
    case EDocumentStatus.FORMAT:
      print('CNPJ\'s format is wrong!');
      break;
    case EDocumentStatus.EMPTY:
      print('CNPJ is empty!');
      break;
    default:
      print('$_cnpj is a valid CNPJ');
  }

  // generate a random cpf
  var _cpf = CnpjCpfBase.randomCpf(withMask: true);
  print('a random CPF: $_cnpj');

  // masking a cpf
  var _fromCpf = '89638567040';
  _cpf = CnpjCpfBase.maskCpf(_fromCpf);
  print('masking a CPF, from this: $_fromCpf to this: $_cpf');
  // it's possible to masking partially
  _fromCpf = '89638567';
  _cpf = CnpjCpfBase.maskCpf(_fromCpf);
  print('masking partially a CPF, from this: $_fromCpf to this: $_cpf');

  // unmasking a cpf
  _fromCpf = '510.867.680-86';
  _cpf = CnpjCpfBase.onlyNumbers(_fromCpf);
  print('unmasking a CPF, from this: $_fromCpf to this: $_cpf');

  // validate a cnpj
  // there are two kinds of validation
  // the first one (the simplest one) just returns a bool value
  if (CnpjCpfBase.isCpfValid(_cpf)) {
    print('$_cpf is a valid CPF');
  }

  // the second one returns a enum that describe the status
  _cpf = '';
  _status = CnpjCpfBase.cpfValidate(_cpf);
  switch (_status) {
    case EDocumentStatus.DIGIT:
      print('CPF\'s digit is wrong!');
      break;
    case EDocumentStatus.FORMAT:
      print('CPF\'s format is wrong!');
      break;
    case EDocumentStatus.EMPTY:
      print('CPF is empty!');
      break;
    default:
      print('$_cpf is a valid CPF');
  }
}
```

## Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

[tracker]: https://github.com/evandrofuhr/cnpj_cpf_helper/issues


CHANGELOG.md
## 1.0.8
- Documentation

## 1.0.7
- Fixed a bug on masking the CPF and tests

## 1.0.6
- Fixed a bug on masking the CPF

## 1.0.5
- Package description improvement (again)

## 1.0.4
- Package description improvement

## 1.0.3
- Remove unused enum

## 1.0.2
- Update the README.md 

## 1.0.1
- Update the Git repository

## 1.0.0
- Initial version, created by Stagehand




test/cnpj_cpf_helper_test.dart
import 'package:test/test.dart';
import 'package:cnpj_cpf_helper/cnpj_cpf_helper.dart';

void main() {
  group('CNPJ tests', () {
    test('CNPJ valid with mask', () {
      expect(CnpjCpfBase.isCnpjValid('07.170.188/0001-65'), isTrue);
    });
    test('CNPJ valid without mask', () {
      expect(CnpjCpfBase.isCnpjValid('07170188000165'), isTrue);
    });

    test('CNPJ invalid with mask', () {
      expect(CnpjCpfBase.isCnpjValid('07.170.188/0001-00'), isFalse);
    });
    test('CNPJ invalid without mask', () {
      expect(CnpjCpfBase.isCnpjValid('07170188000100'), isFalse);
    });

    test('CNPJ masking', () {
      expect(CnpjCpfBase.maskCnpj('07170188000165'),
          matches('\\d{2}.\\d{3}.\\d{3}/\\d{4}-\\d{2}'));
    });
  });

  group('CPF tests', () {
    test('CPF valid with mask', () {
      expect(CnpjCpfBase.isCpfValid('335.742.720-65'), isTrue);
    });
    test('CPF valid without mask', () {
      expect(CnpjCpfBase.isCpfValid('33574272065'), isTrue);
    });

    test('CPF invalid with mask', () {
      expect(CnpjCpfBase.isCpfValid('335.742.720-00'), isFalse);
    });
    test('CPF invalid without mask', () {
      expect(CnpjCpfBase.isCpfValid('33574272000'), isFalse);
    });

    test('CPF masking', () {
      expect(CnpjCpfBase.maskCpf('33574272065'),
          matches('\\d{3}.\\d{3}.\\d{3}-\\d{2}'));
    });
  });
}


lib/src/cnpj_cpf_base.dart
import 'dart:math';

import 'package:cnpj_cpf_helper/cnpj_cpf_helper.dart';

/// Abstract base class to handle validations on CNPJ and CPF documents
abstract class CnpjCpfBase {
  static final _specialChar = '#';

  /// Removes all non-number characters from string
  static String onlyNumbers(String value) {
    return value?.replaceAll(RegExp('[^0-9]'), '')?.trim() ?? '';
  }

  /// Returns if CNPJ document is valid
  static bool isCnpjValid(String value) {
    return cnpjValidate(value) == EDocumentStatus.OK;
  }

  /// Returns if CPF document is valid
  static bool isCpfValid(String value) {
    return cpfValidate(value) == EDocumentStatus.OK;
  }

  /// Returns the CNPJ document Status by the enum `EDocumentStatus`
  static EDocumentStatus cnpjValidate(String value) {
    var _workingValue = onlyNumbers(value);
    if (_workingValue.isEmpty) return EDocumentStatus.EMPTY;
    if (_workingValue.length != 14) return EDocumentStatus.FORMAT;
    var _numbers = _workingValue
        .split('')
        .map<int>((e) => int.parse(e))
        .take(12)
        .toList()
        .reversed
        .toList();
    var _digit1 = _getDigitModule11(_numbers, 2, 9);
    _numbers.insert(0, _digit1);
    var _digit2 = _getDigitModule11(_numbers, 2, 9);
    _numbers.insert(0, _digit2);
    if (_numbers.reversed.join('') != _workingValue) {
      return EDocumentStatus.DIGIT;
    }
    return EDocumentStatus.OK;
  }

  /// Returns the CPF document Status by the enum `EDocumentStatus`
  static EDocumentStatus cpfValidate(String value) {
    var _workingValue = onlyNumbers(value);
    if (_workingValue.isEmpty) return EDocumentStatus.EMPTY;
    if (_workingValue.length != 11) return EDocumentStatus.FORMAT;
    var _numbers =
        _workingValue.split('').map<int>((e) => int.parse(e)).take(9).toList();
    var _digit1 = _getDigitModule11(_numbers, 10, 2);
    _numbers.add(_digit1);
    var _digit2 = _getDigitModule11(_numbers, 11, 2);
    _numbers.add(_digit2);
    if (_numbers.join('') != _workingValue) {
      return EDocumentStatus.DIGIT;
    }
    return EDocumentStatus.OK;
  }

  /// Returns the digit verification using module 11
  static int _getDigitModule11(
      List<int> numbers, int iniFactor, int endFactor) {
    var _factor = iniFactor;
    var _sum = 0;
    for (var _number in numbers) {
      _sum += _number * _factor;
      if (_factor == endFactor) {
        _factor = iniFactor;
      } else {
        if (iniFactor > endFactor) {
          _factor--;
        } else {
          _factor++;
        }
      }
    }
    var _module = _sum % 11;
    var _digit = 11 - _module;
    return _digit < 10 ? _digit : 0;
  }

  /// Generates a random CNPJ. Just for tests
  static String randomCnpj({bool withMask = false}) {
    var _numbers = <int>[];
    _numbers.add(Random.secure().nextInt(8) + 1);
    _numbers.add(0);
    _numbers.add(0);
    _numbers.add(0);
    for (var i = 0; i < 8; i++) {
      _numbers.add(Random.secure().nextInt(9));
    }
    _numbers.insert(0, _getDigitModule11(_numbers, 2, 9));
    _numbers.insert(0, _getDigitModule11(_numbers, 2, 9));
    var _cnpj = _numbers.reversed.join();
    if (withMask) {
      return maskCnpj(_cnpj);
    }
    return _cnpj;
  }

  /// Generates a random CPF. Just for tests
  static String randomCpf({bool withMask = false}) {
    var _numbers = <int>[];
    for (var i = 0; i < 9; i++) {
      _numbers.add(Random.secure().nextInt(9));
    }
    _numbers.add(_getDigitModule11(_numbers, 10, 2));
    _numbers.add(_getDigitModule11(_numbers, 11, 2));
    var _cpf = _numbers.join();
    if (withMask) {
      return maskCpf(_cpf);
    }
    return _cpf;
  }

  /// Returns the given value masquerade like CNPJ
  static String maskCnpj(String value) {
    var _workingValue =
        onlyNumbers(value).padRight(14, _specialChar).substring(0, 14);
    _workingValue = '${_workingValue.substring(0, 2)}.'
        '${_workingValue.substring(2, 5)}.'
        '${_workingValue.substring(5, 8)}/'
        '${_workingValue.substring(8, 12)}-'
        '${_workingValue.substring(12, 14)}';
    return _removeSpecialChar(_workingValue);
  }

  /// Returns the given value masquerade like CPF
  static String maskCpf(String value) {
    var _workingValue =
        onlyNumbers(value).padRight(11, _specialChar).substring(0, 11);
    _workingValue = '${_workingValue.substring(0, 3)}.'
        '${_workingValue.substring(3, 6)}.'
        '${_workingValue.substring(6, 9)}-'
        '${_workingValue.substring(9, 11)}';
    return _removeSpecialChar(_workingValue);
  }

  // Returns the given value without the internal use character
  static String _removeSpecialChar(String value) {
    var _workingValue = value;
    if (_workingValue.contains(_specialChar)) {
      _workingValue =
          _workingValue.substring(0, _workingValue.indexOf(_specialChar));
    }
    return _workingValue;
  }
}


example/cnpj_cpf_helper_example.dart
import 'package:cnpj_cpf_helper/cnpj_cpf_helper.dart';

void main() {
  // generate a random cnpj
  var _cnpj = CnpjCpfBase.randomCnpj(withMask: true);
  print('a random CNPJ: $_cnpj');

  // masking a cnpj
  var _fromCnpj = '26954374000130';
  _cnpj = CnpjCpfBase.maskCnpj(_fromCnpj);
  print('masking a CNPJ, from this: $_fromCnpj to this: $_cnpj');
  // it's possible to masking partially
  _fromCnpj = '2695437';
  _cnpj = CnpjCpfBase.maskCnpj(_fromCnpj);
  print('masking partially a CNPJ, from this: $_fromCnpj to this: $_cnpj');

  // unmasking a cnpj
  _fromCnpj = '03.451.682/0001-47';
  _cnpj = CnpjCpfBase.onlyNumbers(_fromCnpj);
  print('unmasking  a CNPJ, from this: $_fromCnpj to this: $_cnpj');

  // validate a cnpj
  // there are two kinds of validation
  // the first one (the simplest one) just returns a bool value
  if (CnpjCpfBase.isCnpjValid(_cnpj)) {
    print('$_cnpj is a valid CNPJ');
  }

  // the second one returns a enum that describe the status
  _cnpj = '48.608.581/0005-00'; // invalid digit
  var _status = CnpjCpfBase.cnpjValidate(_cnpj);
  switch (_status) {
    case EDocumentStatus.DIGIT:
      print('CNPJ\'s digit is wrong!');
      break;
    case EDocumentStatus.FORMAT:
      print('CNPJ\'s format is wrong!');
      break;
    case EDocumentStatus.EMPTY:
      print('CNPJ is empty!');
      break;
    default:
      print('$_cnpj is a valid CNPJ');
  }

  // generate a random cpf
  var _cpf = CnpjCpfBase.randomCpf(withMask: true);
  print('a random CPF: $_cnpj');

  // masking a cpf
  var _fromCpf = '89638567040';
  _cpf = CnpjCpfBase.maskCpf(_fromCpf);
  print('masking a CPF, from this: $_fromCpf to this: $_cpf');
  // it's possible to masking partially
  _fromCpf = '89638567';
  _cpf = CnpjCpfBase.maskCpf(_fromCpf);
  print('masking partially a CPF, from this: $_fromCpf to this: $_cpf');

  // unmasking a cpf
  _fromCpf = '510.867.680-86';
  _cpf = CnpjCpfBase.onlyNumbers(_fromCpf);
  print('unmasking a CPF, from this: $_fromCpf to this: $_cpf');

  // validate a cnpj
  // there are two kinds of validation
  // the first one (the simplest one) just returns a bool value
  if (CnpjCpfBase.isCpfValid(_cpf)) {
    print('$_cpf is a valid CPF');
  }

  // the second one returns a enum that describe the status
  _cpf = '';
  _status = CnpjCpfBase.cpfValidate(_cpf);
  switch (_status) {
    case EDocumentStatus.DIGIT:
      print('CPF\'s digit is wrong!');
      break;
    case EDocumentStatus.FORMAT:
      print('CPF\'s format is wrong!');
      break;
    case EDocumentStatus.EMPTY:
      print('CPF is empty!');
      break;
    default:
      print('$_cpf is a valid CPF');
  }
}
